kong/plugins/hmac-auth/access.lua

```diff
@@ -0,0 +1,130 @@
+local cache = require "kong.tools.database_cache"
+local stringy = require "stringy"
+local responses = require "kong.tools.responses"
+local constants = require "kong.constants"
+
+local AUTHORIZATION = "authorization"
+local PROXY_AUTHORIZATION = "proxy-authorization"
+local DATE = "date"
+local SIGNATURE_NOT_VALID = "HMAC signature cannot be verified"
+
+local _M = {}
+
+
+local function retrieve_hmac_fields(request, header_name, conf)
+  local username, signature, algorithm
+  local authorization_header = request.get_headers()[header_name]
+  if authorization_header then
+    -- Authentication: hmac username:base64(hmac-sha1(Date))
+    
+    local iterator, iter_err = ngx.re.gmatch(authorization_header, "\\s*[Hh]mac\\s*(.+)")
+    if not iterator then
+      ngx.log(ngx.ERR, iter_err)
+      return
+    end
+
+    local m, err = iterator()
+    if err then
+      ngx.log(ngx.ERR, err)
+      return 
+    end
+    
+    if m and table.getn(m) > 0 then
+      local hmac_fields = stringy.split(m[1], ":")
+      if hmac_fields and #hmac_fields > 2 then
+        username = hmac_fields[1]
+        signature = ngx.decode_base64(hmac_fields[2])
+        algorithm = hmac_fields[3]
+      end  
+    end
+  end
+    
+  if conf.hide_credentials then
+    request.clear_header(header_name)
+  end
+  
+  return username, signature, algorithm
+end
+
+
+local function validate_signature(request, secret, signature, algorithm, defaultClockSkew)
+  -- validate clock skew
+  local date = request.get_headers()[DATE]
+  local requestTime = ngx.parse_http_time(date)  
+  if requestTime == nil then
+    responses.send_HTTP_UNAUTHORIZED(SIGNATURE_NOT_VALID)  
+  end
+  
+  local skew = math.abs(ngx.time() - requestTime)
```

Comment:
ok I ll make that change. But I do not see such optimization in existing plugins. I am basically learning things by looking at existing code.  So it would be very helpful if you guys prepare a doc for standards, do's and don'ts.